Implementation/40907/40908/40904/add duration sort and filter and api

dombesz opened the pull request:
See OP#40908 & OP#40907 & OP#40904 & OP#42723

dombesz committed:
Add duration as a sortable property column

dombesz committed:
Add duration filter for work packages

dombesz committed:
Add duration column to the work package table

dombesz commented on lib/open_project/feature_decisions.rb:
```diff
@@ -31,5 +31,9 @@ module FeatureDecisions
       Setting.feature_storages_module_active
     end
+
+    def self.work_packages_duration_field_active?
```
Please check whether the feature flag I added is correct.

dombesz committed:
Add comment to the OpenProject::FeatureDecisions

ulferts commented on lib/api/v3/work_packages/work_package_representer.rb:
```diff
@@ -391,6 +391,17 @@ def initialize(model, current_user:, embed_links: false)
+                   represented.milestone? || !OpenProject::FeatureDecisions.work_packages_duration_field_active?
+                 },
+                 getter: ->(*) do
+                   datetime_formatter.format_duration_from_hours(represented.duration,
```
There is a mismatch in units here. The duration column works in days since values smaller than a day are not possible. The getter currently treats the duration as if its value would be denoted in hours. The value thus needs to be multiplied by 24 to be correct. 

While not part of this PR, the column looks weird in the UI for a similar reason:

<img width="165" alt="image" src="https://user-images.githubusercontent.com/617519/172385328-9e0b19b3-a55d-4632-b8f4-63b0df0a0168.png">

Even if the value were to be displayed with their correct value, it is't readable to the end user like this. I forwarded that question to product.

dombesz replied:
DONE: b9994104d1143a7676ae74b47c6c286c7edc8245

We can address the day format in a separate PR.

ulferts commented on app/models/queries/work_packages/filter/duration_filter.rb:
```diff
@@ -0,0 +1,42 @@
+      super + " OR #{WorkPackage.table_name}.duration=0"
+    else
+      super
+    end
```
It is not really clear to me what to do with milestone typed work packages. Since they are not supposed to have a duration, as in they do not have the property, they probably should never be returned. This might be different for the `None` operator since then the notion would be 'Work packages without duration' which in my opinion includes milestones. 

Currently this is even harder since milestones still have a duration of 1 as of now.

I am open to suggestions here.

dombesz replied:
DONE: b9994104d1143a7676ae74b47c6c286c7edc8245

It does not return the milestone type wps, when the filter is on, but it does return them, when the filter is "none".

ulferts replied:
👍

ulferts commented on lib/open_project/feature_decisions.rb:
```diff
@@ -28,8 +28,24 @@
+    #     release_date = Date.new(2022,01,01)
+    #     Setting.feature_storages_module_active || Date.today > release_date
+    #   end
+
```
Thanks

ulferts reviewed (changes requested):
Mostly fine except for my remarks inside the code and the two following things:

* The `duration` property is currently not writable which is fine in itself but when trying to update it anyway by sending a PATCH including a duration value, the code runs into a `NoMethodError`. 
* The `duration` property now shows up in the form configuration ("Administration" -> "Work packages" -> "Types" -> "Click on some type in the list" -> "Form configuration"). This happens even though the feature flag disables the "duration" property. Since in 12.3 we have to work towards not allowing the duration column to be removable at all, we can also do it now and never show the property:

<img width="568" alt="image" src="https://user-images.githubusercontent.com/617519/172396764-fa248c2f-a038-4bf9-889c-a0dfe801b123.png">

ulferts commented on lib/api/v3/work_packages/schema/work_package_schema_representer.rb:
```diff
@@ -117,6 +117,18 @@ def initialize(schema, self_link:, **context)
+          # This enables hiding the duration field on the form_configuration
+          ::Type.add_constraint :duration, ->(*) {
+            OpenProject::FeatureDecisions.work_packages_duration_field_active?
+          }
```
We can simply always hide the duration property. Even once we have developed the duration feature to its fullest, will that field be excluded from being managed in the form configuration. This is because of https://community.openproject.org/projects/openproject/work_packages/40539
which will result in the dates and the duration to always be displayed in the work package page. But for 12.2 we can start by excluding the `:duration`.

dombesz replied:
DONE: f73b1412f1eb1be91f4e957e25f99f96d08de206

ulferts reviewed (changes requested):
Only the tiny thing about the form configuration where my first remark was misleading left to be done. All fine otherwise.

dombesz committed:
Remove duration from the work package form configuration

ulferts committed:
Revert "Remove duration from the work package form configuration"

This reverts commit f73b1412f1eb1be91f4e957e25f99f96d08de206.

ulferts committed:
remove duration from form configuration

The property is never to be displayed there since it will receive a fixed place in 12.3

ulferts committed:
decrease MethodABC size

ulferts commented on app/models/type/attributes.rb:
```diff
@@ -32,13 +32,15 @@ module Type::Attributes
                 derived_due_date
-                derived_estimated_time].freeze
+                derived_estimated_time
+                duration].freeze
```
@dombesz apparently `duration` isn't the first property to be excluded from the form configuration. So adding it here got rid of the property.

ulferts commented:
Merging since the failing spec cannot be reproduced locally.
